File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sonix/simplified/MultiTrackSectionSimplified.kt
package com.musicmuni.voxatrace.demo.sonix.simplified

import android.content.Context
import android.widget.Toast
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.musicmuni.voxatrace.sonix.SonixMixer
import com.musicmuni.voxatrace.demo.util.copyAssetToFile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Simplified Multi-Track Section using SonixMixer.create().
 *
 * Demonstrates the zero-config factory pattern for multi-track playback.
 * Compare this to MultiTrackSection.kt which uses SonixMixer.Builder().
 */
@Composable
fun MultiTrackSectionSimplified(context: Context) {
    val scope = rememberCoroutineScope()
    val localContext = LocalContext.current

    var mixer by remember { mutableStateOf<SonixMixer?>(null) }
    var status by remember { mutableStateOf("Loading...") }
    var backingVolume by remember { mutableFloatStateOf(0.8f) }
    var vocalVolume by remember { mutableFloatStateOf(1.0f) }

    // UI state derived from SonixMixer StateFlows
    val isPlaying by mixer?.isPlaying?.collectAsState() ?: remember { mutableStateOf(false) }
    val currentTimeMs by mixer?.currentTime?.collectAsState() ?: remember { mutableStateOf(0L) }
    val error by mixer?.error?.collectAsState() ?: remember { mutableStateOf(null) }
    val durationMs = mixer?.duration ?: 0L

    // Show toast on error
    LaunchedEffect(error) {
        error?.let { err ->
            Toast.makeText(localContext, "Error: ${err.message}", Toast.LENGTH_LONG).show()
        }
    }

    // Initialize mixer
    LaunchedEffect(Unit) {
        // Copy assets to files
        val (backingPath, vocalPath) = withContext(Dispatchers.IO) {
            val backing = copyAssetToFile(context, "sample.m4a")
            val vocal = copyAssetToFile(context, "vocal.m4a")
            Pair(backing.absolutePath, vocal.absolutePath)
        }

        // CREATE MIXER - ZERO-CONFIG FACTORY!
        val newMixer = SonixMixer.create()
        mixer = newMixer

        // Just pass file paths - no manual decoding needed!
        // Wait for both tracks to load before updating status
        val backingSuccess = newMixer.addTrack("backing", backingPath)
        val vocalSuccess = newMixer.addTrack("vocal", vocalPath)

        if (backingSuccess && vocalSuccess) {
            status = "Loaded 2 tracks"
        } else {
            status = "Failed to load tracks"
        }
    }

    // Cleanup
    DisposableEffect(Unit) {
        onDispose { mixer?.release() }
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text("Multi-Track (Simplified API)", style = MaterialTheme.typography.titleMedium)
        Text("Status: $status", style = MaterialTheme.typography.bodySmall)

        // Time display & seek
        Text("${formatTimeMs(currentTimeMs)} / ${formatTimeMs(durationMs)}")
        Slider(
            value = if (durationMs > 0) currentTimeMs.toFloat() / durationMs else 0f,
            onValueChange = { mixer?.seek((it * durationMs).toLong()) },
            enabled = mixer != null,
            modifier = Modifier.fillMaxWidth()
        )

        // Per-track volume controls
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("Backing:", modifier = Modifier.width(70.dp))
            Slider(
                value = backingVolume,
                onValueChange = { backingVolume = it; mixer?.setTrackVolume("backing", it) },
                modifier = Modifier.weight(1f)
            )
            Text("${(backingVolume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("Vocal:", modifier = Modifier.width(70.dp))
            Slider(
                value = vocalVolume,
                onValueChange = { vocalVolume = it; mixer?.setTrackVolume("vocal", it) },
                modifier = Modifier.weight(1f)
            )
            Text("${(vocalVolume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        // Playback controls
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Button(onClick = { mixer?.play() }, enabled = mixer != null && !isPlaying) { Text("▶") }
            Button(onClick = { mixer?.pause() }, enabled = isPlaying) { Text("⏸") }
            Button(onClick = { mixer?.stop() }, enabled = mixer != null) { Text("⏹") }
        }

        // Fade buttons for vocal track
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Button(
                onClick = { mixer?.fadeTrackVolume("vocal", 0.2f, 500); vocalVolume = 0.2f },
                enabled = mixer != null
            ) { Text("Fade Vocal ↓") }
            Button(
                onClick = { mixer?.fadeTrackVolume("vocal", 1.0f, 500); vocalVolume = 1.0f },
                enabled = mixer != null
            ) { Text("Fade Vocal ↑") }
        }
    }
}

private fun formatTimeMs(ms: Long): String {
    val seconds = (ms / 1000) % 60
    val minutes = (ms / 1000) / 60
    return String.format("%d:%02d", minutes, seconds)
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Sonix/MultiTrackSectionSimplified.swift
import SwiftUI
import VoxaTrace

/// Simplified Multi-Track Section using SonixMixer.create().
///
/// Demonstrates the zero-config factory pattern for multi-track playback.
/// Uses type-safe observers (no force casts like `as! Bool`).
/// Compare this to MultiTrackSection.swift which uses SonixMixerConfig.Builder() + SonixMixer.create().
struct MultiTrackSectionSimplified: View {
    @State private var mixer: SonixMixer?
    @State private var isPlaying = false
    @State private var currentTimeMs: Int64 = 0
    @State private var durationMs: Int64 = 0
    @State private var backingVolume: Float = 0.8
    @State private var vocalVolume: Float = 1.0
    @State private var status = "Loading..."
    @State private var isLoaded = false

    // Task handles for cancellation
    @State private var playingTask: Task<Void, Never>?
    @State private var timeTask: Task<Void, Never>?
    @State private var errorTask: Task<Void, Never>?

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text("Multi-Track (Simplified API)")
                    .font(.headline)
                Spacer()
                if !isLoaded {
                    ProgressView()
                        .scaleEffect(0.8)
                }
            }

            Text("Status: \(status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // Time display
            Text("\(formatTime(currentTimeMs)) / \(formatTime(durationMs))")
                .font(.body)

            // Seek slider
            Slider(value: Binding(
                get: { durationMs > 0 ? Double(currentTimeMs) / Double(durationMs) : 0 },
                set: { newValue in
                    let seekPos = Int64(newValue * Double(durationMs))
                    mixer?.seek(positionMs: seekPos)
                }
            ))
            .disabled(!isLoaded)

            // Backing volume
            HStack {
                Text("Backing:")
                    .frame(width: 70, alignment: .leading)
                Slider(value: Binding(
                    get: { Double(backingVolume) },
                    set: { newValue in
                        backingVolume = Float(newValue)
                        mixer?.setTrackVolume(name: "backing", volume: backingVolume)
                    }
                ))
                Text("\(Int(backingVolume * 100))%")
                    .frame(width: 40)
            }

            // Vocal volume
            HStack {
                Text("Vocal:")
                    .frame(width: 70, alignment: .leading)
                Slider(value: Binding(
                    get: { Double(vocalVolume) },
                    set: { newValue in
                        vocalVolume = Float(newValue)
                        mixer?.setTrackVolume(name: "vocal", volume: vocalVolume)
                    }
                ))
                Text("\(Int(vocalVolume * 100))%")
                    .frame(width: 40)
            }

            // Playback controls
            HStack(spacing: 12) {
                Button("Play") {
                    mixer?.play()
                }
                .disabled(!isLoaded || isPlaying)
                .buttonStyle(.borderedProminent)

                Button("Pause") {
                    mixer?.pause()
                }
                .disabled(!isPlaying)
                .buttonStyle(.bordered)

                Button("Stop") {
                    mixer?.stop()
                    currentTimeMs = 0
                }
                .disabled(!isLoaded)
                .buttonStyle(.bordered)
            }

            // Fade buttons
            HStack(spacing: 12) {
                Button("Fade Vocal \u{2193}") {
                    mixer?.fadeTrackVolume(name: "vocal", targetVolume: 0.2, durationMs: 500)
                    vocalVolume = 0.2
                }
                .disabled(!isLoaded)
                .buttonStyle(.bordered)
                .controlSize(.small)

                Button("Fade Vocal \u{2191}") {
                    mixer?.fadeTrackVolume(name: "vocal", targetVolume: 1.0, durationMs: 500)
                    vocalVolume = 1.0
                }
                .disabled(!isLoaded)
                .buttonStyle(.bordered)
                .controlSize(.small)
            }
        }
        .task {
            await initializeMixer()
        }
        .onDisappear {
            // Cancel all observer tasks
            playingTask?.cancel()
            timeTask?.cancel()
            errorTask?.cancel()
            mixer?.release()
        }
    }

    private func initializeMixer() async {
        // Copy assets to files
        guard let backingPath = copyAssetToFile(name: "sample", ext: "m4a"),
              let vocalPath = copyAssetToFile(name: "vocal", ext: "m4a") else {
            await MainActor.run {
                status = "Assets not found"
            }
            return
        }

        // CREATE MIXER - ZERO-CONFIG FACTORY!
        let newMixer = SonixMixer.create()

        // Add tracks - just pass file paths, no manual decoding needed!
        let backingSuccess = try? await newMixer.addTrack(name: "backing", filePath: backingPath)
        let vocalSuccess = try? await newMixer.addTrack(name: "vocal", filePath: vocalPath)

        await MainActor.run {
            mixer = newMixer
            if backingSuccess == true && vocalSuccess == true {
                durationMs = newMixer.duration
                isLoaded = true
                status = "Loaded 2 tracks"

                // Type-safe observers - no force casts needed!
                playingTask = newMixer.observeIsPlaying { playing in
                    self.isPlaying = playing
                }
                timeTask = newMixer.observeCurrentTime { time in
                    self.currentTimeMs = time
                }
                errorTask = newMixer.observeError { error in
                    if let error = error {
                        self.status = "Error: \(error.message)"
                    }
                }
            } else {
                status = "Failed to load tracks"
            }
        }
    }

    private func formatTime(_ ms: Int64) -> String {
        let seconds = (ms / 1000) % 60
        let minutes = (ms / 1000) / 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}

#Preview {
    MultiTrackSectionSimplified()
        .padding()
}
